core/commands/UserUpdateCommand.php

```diff
@@ -0,0 +1,345 @@
+				    ( !is_null( $this->access_level ) && $this->access_level < $t_admin_threshold ) ) {
+						throw new ClientException( "Disabling or reducing access level of last admin not allowed.", ERROR_USER_CHANGE_LAST_ADMIN );
+				}
+			}
+		}
+
+		$this->old_user = array(
+			'id' => $this->user_id,
+			'username' => $t_old_username,
+			'real_name' => $t_old_realname,
+			'email' => $t_old_email,
+			'access_level' => $t_old_access_level,
+			'enabled' => $t_old_enabled,
+			'protected' => $t_old_protected
+		);
+
+		$this->new_user = array(
+			'id' => $this->user_id,
+			'username' => $t_new_username ?: $t_old_username,
+			'real_name' => !is_null( $t_new_realname ) ? $t_new_realname : $t_old_realname,
+			'email' => $t_new_email ?: $t_old_email,
+			'access_level' => $t_new_access_level ?: $t_old_access_level,
+			'enabled' => !is_null( $t_new_enabled ) ? $t_new_enabled : $t_old_enabled,
+			'protected' => !is_null( $t_new_protected ) ? $t_new_protected : $t_old_protected
+		);
+	}
+
+	/**
+	 * Process the command.
+	 *
+	 * @return array Command response
+	 */
+	protected function process() {
+		$this->update_user( $this->new_user );
+
+		# Project specific access rights override global levels, hence, for users who are changed
+		# to be administrators, we have to remove project specific rights.
+		if( !is_null( $this->access_level ) ) {
+			$t_admin_threshold = config_get_global( 'admin_site_threshold' );
+			if( ( $this->access_level >= $t_admin_threshold ) && !user_is_administrator( $this->user_id ) ) {
+				user_delete_project_specific_access_levels( $this->user_id );
+			}
+		}
+
+		if( $this->notify_user ) {
+			email_user_changed( $this->user_id, $this->old_user, $this->new_user );
+		}
+
+		event_signal( 'EVENT_MANAGE_USER_UPDATE', array( $this->user_id ) );
+
+		user_clear_cache( $this->user_id );
+		$t_select = array( 'id', 'name', 'real_name', 'email', 'access_level', 'enabled', 'protected' );
+		$t_user = mci_user_get( $this->user_id, $t_select );
+		return array( 'user' => $t_user );
+	}
+
+	/**
+	 * Update user in database.
+	 *
+	 * There was no function to update user before. The functionality was done
+	 * directly in the action page. Given that the database update is done here
+	 * directly to discourage direct usage of the function without validation,
+	 * authorization, triggering of events, etc.
+	 *
+	 * @param array $p_user User data
+	 * @return void
+	 */
+	private function update_user( $p_user ) {
+		db_param_push();
+
+		$t_query = 'UPDATE {user}
+			SET username=' . db_param() . ', email=' . db_param() . ',
+				access_level=' . db_param() . ', enabled=' . db_param() . ',
+				protected=' . db_param() . ', realname=' . db_param() . '
+			WHERE id=' . db_param();
+
+		$t_query_params = array(
+			$p_user['username'],
+			$p_user['email'],
+			$p_user['access_level'],
+			$p_user['enabled'],
+			$p_user['protected'],
+			$p_user['real_name'],
+			$p_user['id'] );
+
+		db_query( $t_query, $t_query_params );
```

Comment:
Shouldn't this be a low-level API (user_update() function in user_api.php) ?

I suggest to use a DbQuery object here instead of using legacy db api calls. See project_update() for example: 
https://github.com/mantisbt/mantisbt/blob/8c5b133eb1058c6222f4f26b44649a65a8d213b3/core/project_api.php#L469-L488

Reply:
We don't have the concept of a low level private/internal API, so will leave here for now. The code is also moved from the action page IIRC. Given that this implementation is tested already, I'll leave as is for now.